cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/repo_auth.go

```diff
@@ -0,0 +1,523 @@
+			}
+		}
+		return updatedSecret, true, true, nil
+	}
+}
+
+func (s *Server) validateUserManagedRepoSecret(
+	ctx context.Context,
+	repoName types.NamespacedName,
+	repoType string,
+	tlsConfig *corev1.PackageRepositoryTlsConfig,
+	auth *corev1.PackageRepositoryAuth) (*apiv1.Secret, error) {
+	var secretRefTls, secretRefAuth string
+	if tlsConfig != nil {
+		if tlsConfig.GetCertAuthority() != "" {
+			return nil, status.Errorf(codes.InvalidArgument, "Secret Ref must be used with user managed secrets")
+		} else if tlsConfig.GetSecretRef().GetName() != "" {
+			secretRefTls = tlsConfig.GetSecretRef().GetName()
+		}
+	}
+
+	if auth != nil {
+		if auth.GetDockerCreds() != nil ||
+			auth.GetHeader() != "" ||
+			auth.GetTlsCertKey() != nil ||
+			auth.GetUsernamePassword() != nil {
+			return nil, status.Errorf(codes.InvalidArgument, "Secret Ref must be used with user managed secrets")
+		} else if auth.GetSecretRef().GetName() != "" {
+			secretRefAuth = auth.GetSecretRef().GetName()
+		}
+	}
+
+	var secretRef string
+	if secretRefTls != "" && secretRefAuth != "" && secretRefTls != secretRefAuth {
+		// flux repo spec only allows one secret per HelmRepository CRD
+		return nil, status.Errorf(
+			codes.InvalidArgument, "TLS config secret and Auth secret must be the same")
+	} else if secretRefTls != "" {
+		secretRef = secretRefTls
+	} else if secretRefAuth != "" {
+		secretRef = secretRefAuth
+	}
+
+	var secret *apiv1.Secret
+	if secretRef != "" {
+		// check that the specified secret exists
+		if typedClient, err := s.clientGetter.Typed(ctx, s.kubeappsCluster); err != nil {
+			return nil, err
+		} else if secret, err = typedClient.CoreV1().Secrets(repoName.Namespace).Get(ctx, secretRef, metav1.GetOptions{}); err != nil {
+			return nil, statuserror.FromK8sError("get", "secret", secretRef, err)
+		} else {
+			// also check that the data in the opaque secret corresponds
+			// to specified auth type, e.g. if AuthType is
+			// PackageRepositoryAuth_PACKAGE_REPOSITORY_AUTH_TYPE_BASIC_AUTH,
+			// check that the secret has "username" and "password" fields, etc.
+			// it appears flux does not care about the k8s secret type (opaque vs tls vs basic-auth, etc.)
+			// https://github.com/fluxcd/source-controller/blob/bc5a47e821562b1c4f9731acd929b8d9bd23b3a8/controllers/helmrepository_controller.go#L357
+			if secretRefTls != "" && secret.Data["caFile"] == nil {
+				return nil, status.Errorf(codes.Internal, "Specified secret [%s] missing field 'caFile'", secretRef)
+			}
+			if secretRefAuth != "" {
+				switch auth.Type {
+				case corev1.PackageRepositoryAuth_PACKAGE_REPOSITORY_AUTH_TYPE_BASIC_AUTH:
+					if secret.Data["username"] == nil || secret.Data["password"] == nil {
+						return nil, status.Errorf(codes.Internal, "Specified secret [%s] missing fields 'username' and/or 'password'", secretRef)
+					}
+				case corev1.PackageRepositoryAuth_PACKAGE_REPOSITORY_AUTH_TYPE_TLS:
```

Comment:
according to the flux doc, TLS is not yet supported by OCI (https://fluxcd.io/flux/components/source/helmrepositories/#tls-authentication).
should the code check for the repo type (if indeed flux does not support tls)?

Reply:
OK, will add this check. In general, I don't like to have that because literally tomorrow flux might support it and then I would need to change my code again to get rid of this check